plugins/net_plugin/include/eosio/net_plugin/auto_bp_peering.hpp

```diff
@@ -326,17 +404,29 @@ class bp_connection_manager {
       auto& idx = gossip_bps.index.get<by_producer>();
       bool diff = false;
       for (const auto& peer : msg.peers) {
-         if (auto i = idx.find(boost::make_tuple(peer.producer_name, boost::cref(peer.server_address))); i != idx.end()) {
-            if (i->sig != peer.sig) { // signature has changed, producer_name and server_address has not changed
+         if (auto i = idx.find(boost::make_tuple(peer.producer_name, boost::cref(peer.server_endpoint))); i != idx.end()) {
+            if (i->sig != peer.sig && peer.expiration >= i->expiration) { // signature has changed, producer_name and server_endpoint has not changed
                gossip_bps.index.modify(i, [&peer](auto& m) {
-                  m.sig = peer.sig; // update the signature, producer_name and server_address has not changed
+                  m.outbound_ip_address = peer.outbound_ip_address;
+                  m.expiration = peer.expiration;
+                  m.sig = peer.sig;
                });
                diff = true;
             }
          } else {
-            if (idx.count(peer.producer_name) >= max_bp_peers_per_producer) {
-               // only allow max_bp_peers_per_producer, choose one to remove
-               gossip_bps.index.erase(idx.find(peer.producer_name));
+            auto r = idx.equal_range(peer.producer_name);
+            if (std::distance(r.first, r.second) >= max_bp_gossip_peers_per_producer) {
+               // remove entry with min expiration
+               auto min_expiration_itr = r.first;
+               auto min_expiration = min_expiration_itr->expiration;
+               ++r.first;
+               for (; r.first != r.second; ++r.first) {
+                  if (r.first->expiration < min_expiration) {
+                     min_expiration = r.first->expiration;
+                     min_expiration_itr = r.first;
+                  }
+               }
+               gossip_bps.index.erase(min_expiration_itr);
```

Comment:
Suppose one node changes his `peer.server_endpoint` once. We will have two entries in our index, one with the old endpoint and one with the new one. 
When will we ever remove the entry with the old endpoint? Will we ever have `std::distance(r.first, r.second) >= max_bp_gossip_peers_per_producer`?

Reply:
It will be expired in `expire_gossip_bp_peers()`. Until it expires we will have both.

Reply:
Why have this code here then? We could just do the `gossip_bps.index.insert(peer);` here, and rely on `expire_gossip_bp_peers()`  for retiring the expired ones.

Reply:
It is considered invalid to gossip more than 8 bp entries. We remove here so we remain <= 8.

Reply:
Sure, but these are going to be removed anyways when they expire, won't they? Is there a reason to do it immediately at this point?

Reply:
We need to remove it here or when we send out the bp gossip msgs otherwise other nodes will reject the bp gossip msg.